extension_distribution.md

```diff
@@ -0,0 +1,296 @@
+# Extension Distribution Spec
+
+## Problem
+
+Tilt's current extension system is optimized to support libraries of
+Tiltfile/Starlark functions.
+
+Rather than copy-and-pasting snippets of Tiltfiles, you load
+those snippets from a common location. To use an extension, the Tiltfile
+you're working on needs to manually load the extension.
+
+We know that most extensions get very low usage. The ones that do get a lot of
+usage are linked directly from the main Tilt docs.
+
+How can we better encourage an ecosystem where users discover and distribute
+more sophisticated extensions that interact with the Tilt API?
```

Comment:
I dunno if you want to get into it here, but I've been trying to figure out why global extensions before per-project extensions:
### Per-project extensions (i.e., the Tiltfile specifies the project uses the extension and can rely on it running):
1. One user can enable it in a project and their whole team benefits
2. If a project can reference an extension, then it can provide functionality using that extension (e.g., declare a resource that uses a framework enabled by an extension, or a resource that's triggered when another resource finishes)
3. Feels like it's strictly less work than global extensions? (don't need to deal with the global persistent api objects)
### Global extensions:
1. (I think?) only make sense for generic functionality that knows little about the project being run (like, you could have a global extension that honks on failure, but an extension that ignores a specific spammy container feels like it shouldn't be global)
2. Create more room for non-hermetic build "works on my machine" issues

IIUC:
1. The work for per-project extensions is mostly a subset of the work for global extensions
2. per-project extensions seem likely to be easier to spread, since only one person per project has to enable them
3. my gut is that there are many more use cases for per-project extensions than global extensions

Have we considered doing per-project extensions as a stepping stone on the way to global extensions? Is it worth discussing what motivates this decision up here? Or maybe discussing per-project extensions under future work?

Reply:
that's a deep question!

I think it speaks more to product questions rather than engineering questions, so i'll defer to @ellenkorbes here

A couple non-conclusive observations though:

1) I do think Tilt has a positioning problem in that some people see it complementing an IDE, and some people see it as complementing a build system. I think the global vs project extension question speaks to that.

2) We already have global reconcilers (i.e., all the reconcilers under https://github.com/tilt-dev/tilt/tree/master/internal/controllers/core), and I think there's an implicit product goal that any of those reconcilers could be extensions.

Reply:
> I do think Tilt has a positioning problem in that some people see it complementing an IDE, and some people see it as complementing a build system. I think the global vs project extension question speaks to that.

That framing seems really useful for a lot of conversations!

> We already have global reconcilers (i.e., all the reconcilers under https://github.com/tilt-dev/tilt/tree/master/internal/controllers/core), and I think there's an implicit product goal that any of those reconcilers could be extensions.

I think my main concern in that vein isn't whether the reconcilers can be extensions, but rather ensuring that once they are, a project is able to rely on those extensions being loaded (like, if you call k8s_resource but don't have the kdisco controller, you might be a little sad).

Reply:
> my gut is that there are many more use cases for per-project extensions than global extensions
> ensuring that once they are, a project is able to rely on those extensions being loaded

Reply:
> my gut is that there are many more use cases for per-project extensions than global extensions

> ensuring that once they are, a project is able to rely on those extensions being loaded

My perception is that global extensions generally relate to someone's personal settings on their dev environment, not to the application itself, similarly to customizing an IDE i.e. display things a certain way, automate a certain process, always show a certain custom button, etc.

Meanwhile, project-level extensions affect how the application works i.e. integration with a package manager or build system or something that watches a certain kind of resource, etc.

In addition, extensions that affect the application itself, whether global or project-specific, usually need specific Tiltfile calls to pass in parameters and so on, and that would imply `load()` calls that in turn can be configured to ensure that everything that needs to be present, is.

@nicks @landism Does this make sense?

Reply:
> Does this make sense?

Yep, that's how I understand the breakdown.

My comment was mostly about their relative priority:
1) trying to communicate that my gut is that project-level extensions are both more useful and cheaper than global extensions
2) seeing if I could articulate that well enough to convince others, or else learn what I'm missing that led us to do global extensions first
3) otherwise, wondering if/when we might tackle project-level extensions

Reply:
I feel like from a usability standpoint it's probably more convenient to distribute/store them all globally. This way UI-style extensions that should "magically appear" in all Tilt instances can do that, and the ones that need to be specifically called with custom parameters can be invoked from the Tiltfile as usual.

You're right that we should have some sort of verification mechanism in place, so that we don't try to start things up while there are missing load bearing extensions, so to speak, which Tilt should alert the user about, or download automatically, or something like that.

Either way it appears we have some unresolved ambiguities here; I'll book a quick chat so we can all discuss this.

Reply:
fwiw, as I've been playing around with this, I think the global extension vs project extension distinction is kind of a red herring. ultimately, you'll need both.

The more interesting distinction is whether extensions/Tiltfiles can be a bit more modular, and able to load/unload in any project without a lot of mental overhead (sort of like how vscode makes it easy to install extensions and never think about them again).

The current extension system puts the emphasis on loading symbols, and trying to get away from that emphasis